docs/rfc/rfc-003-performance-questions.md

```diff
@@ -0,0 +1,193 @@
+The listed number of nodes a user reported as causing issue was in the thousands.
+We don't currently have evidence about what the upper-limit of nodes that Tendermint's
+P2P stack can scale to.
+
+We need to more concretely understand the source of issue and determine what layer
+is causing a problem. It's possible that the P2P layer, in the absence of any reactors
+sending data, is perfectly capable of managing thousands of peer connections. For
+a reasonable networking and application setup, thousands of connections should not present any
+issue for the application.
+
+We need more data to understand the problem directly. We want to drive the popularity
+and adoption of Tendermint and this will mean allowing for chains with more validators.
+We should follow up with users experiencing this issue. We may then want to benchmark
+the P2P layer in a few simple ways:
+1. By starting a network with 1000 nodes and sending no messages to determine memory
+and CPU consumption in the presence of 1000 nodes.
+2. By testing two nodes connected to each other via P2P, and attempting to
+generate as much load between the two nodes as possible. This will help inform
+how much load is incurred from message sends in an otherwise unencumbered P2P stack.
+
+### Sync
+
+#### Block Syncing is slow
+
+Bootstrapping a new node in a network to the height of the rest of the network
+takes longer than users would like. Block sync requires fetching all of the blocks from
+peers and placing them into the local disk for storage. A useful line of inquiry
+is understanding how quickly a perfectly tuned system _could_ fetch all of the state
+over a network so that we understand how much overhead Tendermint actually adds.
+
+The operation is likely to be _incredibly_ dependent on the environment in which
+the node is being run. The factors that will influence syncing include:
+1. Number of peers that a syncing node may fetch from.
+2. Speed of the disk that a validator is writing to.
+3. Speed of the network connection between the different peers that node is
+syncing from.
+
+We should calculate how quickly this operation _could possibly_ complete for common chains and nodes.
+To calculate how quickly this operation could possibly complete, we should assume that
+a node is reading at line-rate of the NIC and writing at the full drive speed to its
+local storage.
+
+Comparing this theoretical upper-limit to the actual sync times observed by node operators will
+give us a good point of comparison for understanding how much room for improvement there is
+within Tendermint.
+
+### Application
+
+Applications performing complex state transitions have the potential to bottleneck
+the Tendermint node.
+
+#### ABCI block delivery could cause slowdown
+
+ABCI delivers blocks in several methods: `BeginBlock`, `DeliverTx`, `EndBlock`, `Commit`.
+
+Tendermint delivers transactions one-by-one via the `DeliverTx` call. Most of the 
```

Comment:
If I have understood the design of ABCI++ correctly, some of these concerns should be easier to address there.